api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto

```diff
@@ -820,6 +820,13 @@ message HttpConnectionManager {
   // This should be set to `false` in cases where Envoy's view of the downstream address may not correspond to the
   // actual client address, for example, if there's another proxy in front of the Envoy.
   google.protobuf.BoolValue add_proxy_protocol_connection_state = 53;
+
+  // Configurations for RBAC-based authorization audit loggers.
+  // Any RBAC filter in the HTTP filter chain will invoke all the audit loggers configured here if the audit
+  // condition specified in the RBAC is met.
+  //
+  // [#not-implemented-hide:]
+  repeated config.core.v3.TypedExtensionConfig audit_log = 54;
```

Comment:
Do we need a new extension category or is the access log good enough? We really just need `log()` callback, right?

Reply:
The idea to have a new extension with different types of loggers. Here are the [slides](http://go/grpc-auditing-concept-slides) and WIP [doc](http://go/grpc-audit-logging) that you should have access to.

Reply:
cc @rockspore Could you make te slides and doc public then most people could access it? Thanks.

Reply:
Sorry, it's not clear to me whether I could publicize those. But I will link a gRFC PR with most of the context available. Thanks!

Reply:
The other alternative here is that we extend access log configuration to allow different loggers at different request lifecycle points, and they get to access the audit metadata as needed there. That seems less hackish. Given this will almost certaily be used in Envoy once we introduce (there is a clear need IMHO), I think we want to spend time to get this API right CC @envoyproxy/api-shepherds

Reply:
If this is to be done by extending the access log, it means gRPC will have to support the access log entirely (including how it is currently used), right? I am not sure whether this is what they want.

Reply:
RBAC filter publishes its decision and policy name as part of %RESPONSE_CODE_DETAILS%. It's not structured so it's not easy to write a filter over it, but the context is already available to the standard access log.

Reply:
I am not yet familiar with the impl details of Envoy. Could you elaborate on "write a filter over it"? I was imagining the RBAC engine would have handles of the loggers to invoke when it needed to.

IIUC, "decision and policy name" would only be available if the request is denied by the RBAC filter. Otherwise any upstream response would overwrite that piece of information.

The current access log won't fit well since it's not invoked till the entire request/response flow finishes, whereas audit logging should typically happen right after the authorization decision is made. Of course, as @htuch suggested, it would work if access loggers were configured to run at different stages.

Reply:
I think with @htuch and you that the missing piece in Envoy is the "audit context". What's there (shadow policy metadata, response code details) appears to be insufficient. There's also LOG action here which seems to be trying to do the same in yet another way.

Reply:
`LOG` action is very different from what we want. It tells the policy to not enforce anything but simply provides a hint when the access log kicks in.

Reply:
I agree that we want to make sure we get this API right before we do it -- and all of this discussion is valuable!

I am hesitant to support audit logs via the existing access log mechanism, because access logs don't actually make sense for gRPC in the general case due to streaming or other long-lived RPCs.  The normal access log paradigm is that there's one log line for each request that includes both request and response info.  But if an individual RPC lives for hours or days, you wind up having to choose between logging an entry at the start of the RPC that does not include response info or waiting until the RPC is complete to log anything (which would defeat the purpose of using this for audit logs).  And things get even more complicated when you consider cases of multiple messages in either direction on a stream -- it's not clear what should be considered "the request" or "the response" in those cases.  Because of this significant paradigm mismatch, I really don't want to support the xDS access logging extension in gRPC.  But we do need a way to do audit logging, which implies that we need a separate mechanism for that.

What I'd really like to see here is a new, simple extension that is always invoked at the start of a stream and that exposes only initial metadata and authz decision to the logger.

Reply:
Would we really want a new API to cover both audit logging and other generalized logging scenarios? From my perspective, audit logging is more about auditing accesses to specific services so it's very much tied to authn/z and cares little about anything after the stream starts.

@htuch When you said there was a clear need in Envoy, did you mean Envoy also primarily needed this audit logging solution or it was referring to the generalized logging where a logger can access different lifecycle points? I'd love to know more about the use cases for the latter.

Reply:
I think for audit logging specifically, it would be better to have a mechanism specifically for that purpose than to use a more generalized logging mechanism.  For audit logging, it's important to be able to easily validate from looking at the config that the logging will actually be done, and the more complicated that configuration is, the harder it is to do that validation.

To be clear, I am not opposed to having a more general logging solution in the future.  But I think that should be a separate effort from audit logging, even if audit logging could also be implemented as a special case of the more generalized solution.

I think this PR is basically what we want for audit logging.

Reply:
I don't have a very strong opinion on this one way or the other, though if pushed I would tend to agree it would be better to have a common logging API that could be hooked in various different ways and at various lifecycle points.

> The normal access log paradigm is that there's one log line for each request that includes both request and response info. 

I agree this is generally true historically, but we already support in Envoy emitting multiple logs for a connection/stream. See the recent https://github.com/envoyproxy/envoy/pull/25824 for example. 

It's also supported by the TCP proxy: https://github.com/envoyproxy/envoy/blob/3c26c22671587c8918a497961107d16dfcb2747f/api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto#L215-L221

It seems like we might want something similar where we could configure RBAC to just invoke a set of loggers (with the common logging API) when the RBAC decision happens. I think this would be more flexible than a new type of logger.